src/backend/distributed/utils/foreign_key_relationship.c

```diff
@@ -323,15 +326,33 @@ CreateForeignConstraintRelationshipGraph()
 
 	ClearForeignConstraintRelationshipGraphContext();
 
-	MemoryContext fConstraintRelationshipMemoryContext = AllocSetContextCreateInternal(
-		CacheMemoryContext,
-		"Forign Constraint Relationship Graph Context",
-		ALLOCSET_DEFAULT_MINSIZE,
-		ALLOCSET_DEFAULT_INITSIZE,
-		ALLOCSET_DEFAULT_MAXSIZE);
+	/*
+	 * Lazily create our memory context once and reset on every reuse.
+	 * Since we have cleared and invalidated the fConstraintRelationshipGraph, right
+	 * before we can simply reset the context if it was already existing.
+	 */
+	if (ForeignConstraintRelationshipMemoryContext == NULL)
+	{
+		/* make sure we've initialized CacheMemoryContext */
+		if (CacheMemoryContext == NULL)
+		{
+			CreateCacheMemoryContext();
+		}
+
+		ForeignConstraintRelationshipMemoryContext = AllocSetContextCreate(
+			CacheMemoryContext,
+			"Foreign Constraint Relationship Graph Context",
+			ALLOCSET_DEFAULT_MINSIZE,
+			ALLOCSET_DEFAULT_INITSIZE,
+			ALLOCSET_DEFAULT_MAXSIZE);
+	}
+	else
+	{
+		MemoryContextReset(ForeignConstraintRelationshipMemoryContext);
+	}
 
 	MemoryContext oldContext = MemoryContextSwitchTo(
-		fConstraintRelationshipMemoryContext);
+		ForeignConstraintRelationshipMemoryContext);
 
 	fConstraintRelationshipGraph = (ForeignConstraintRelationshipGraph *) palloc(
 		sizeof(ForeignConstraintRelationshipGraph));
```

Comment:
Resetting ForeignConstraintRelationshipMemoryContext would invalidate fConstraintRelationshipGraph as well and then the time IsForeignConstraintRelationshipGraphValid would be accessing a freed memory?

Reply:
https://github.com/citusdata/citus/pull/7236/files#diff-e0be0392562ddabab1d64b436a7cd8225ca5678d64b80c6f01c2be615a3848a8R327 Should have set that to `NULL` already no? I could add an assert to verify and make that case clear I guess

Reply:
>  Should have set that to NULL already no? 

Right. And when looking into that function, I realized that hash_destroy() already deletes the memory context. I know it's more efficient to reset a memory context rather than deleting it but wondering how we were leaking the memory before then.

Reply:
When creating a hashmap with the `HASH_CONTEXT` flag, like we do, the hashmap has an internal context as a child of its current context:
https://github.com/postgres/postgres/blob/ffb69b23115a1ca1d81f7e273d50b75154ae7b0b/src/backend/utils/hash/dynahash.c#L385-L387

This context is then stored on the hashmap via
https://github.com/postgres/postgres/blob/ffb69b23115a1ca1d81f7e273d50b75154ae7b0b/src/backend/utils/hash/dynahash.c#L501-L505

Which is then destroyed when `hash_destroy` deletes its context.

The context we create is the one that is leaking.